### ISSUE TITLE
Integrate with time stepping codes

### ISSUE THREAD

**Issue by dlfivefifty** [JuliaApproximation/SpectralTimeStepping.jl#1]
This should be a home for integrating ApproxFun with standard time-stepping codes, e.g.:

https://github.com/JuliaDiffEq/ODE.jl
https://github.com/luchr/ODEInterface.jl

This follows up on conversations at TU Munich with @luchr and @al-Khwarizmi, and a discourse discussion

https://discourse.julialang.org/t/pros-cons-of-approxfun/396/10

with @ChrisRackauckas.

**Comment by ChrisRackauckas** [comment]
Note that if you just target the DiffEq common interface, the same code will work for DifferentialEquations.jl, Sundials.jl, ODEInterface.jl, and ODE.jl. That's why I'm suggesting doing that.

**Comment by ChrisRackauckas** [comment]
All that would be required is that you'd have to be able to build a Problem type

https://juliadiffeq.github.io/DiffEqDocs.jl/latest/types/ode_types.html

If there's a function that builds the problem from an ApproxFun function, then it should all just work.

**Comment by dlfivefifty** [comment]
Ah OK.   Note that the code here is derelict, and is of the variety of "discretize in time, ∞-dimensional in space" where your previous request was the more traditional pseudo-spectral approach of "∞-dimensional in time, discretize in space".  

I think the latter approach of "∞-dimensional in time, discretize in space" is the right way to go, as it means nothing special has to be done to DiffEq.

**Comment by dlfivefifty** [comment]
For periodic problems it should be really easy.

How do you support constraints?  For example, with a Dirichlet heat equation

$$u_t = u_{xx}, u(t,-1) = u(t,1) = 0$$

It's necessary that we constrain the value at ±1.

**Comment by dlfivefifty** [comment]
Also, is it necessary that `u0` is an AbstractArray?  Maybe a more elegant approach is to allow `u0` to be a `Fun`, and as long as the time stepper is written in Julia, it should work.

**Comment by ChrisRackauckas** [comment]
> Also, is it necessary that u0 is an AbstractArray?

That can be relaxed. Will a `Fun` index like an array?

**Comment by dlfivefifty** [comment]
No.  Why would you need indexing like an array?

**Comment by dlfivefifty** [comment]
Let me rephrase that. a vector-valued Fun behaves like a vector:

``` julia
f=Fun(x->[exp(x);sin(x)],[-1,1])
f[1]   # is Fun(exp,[-1,1])
f[2]  # is Fun(sin,[-1,1])
```

A scalar-valued Fun doesn't at the moment support indexing, but `f[1]` could return `f`.

**Comment by ChrisRackauckas** [comment]
For the update steps which do a linear combination on the partial steps, like in a Runge-Kutta method. For example, say we took two partial RK steps and stored them in `k[1]` and `k[2]`, the next step would compute something like:

``` julia
for i in eachindex(u)
  tmp[i] = a31*k[1][i] + a32*k[2][i]
end
```

Is it better than to treat a `Fun` like a scalar?

**Comment by ChrisRackauckas** [comment]
Well here's the big question: is a `Fun` mutable? Is it something that you'd update in-place?

**Comment by dlfivefifty** [comment]
Yes, a `Fun` should be treated like a scalar.  As I understand it, the following code should work to solve heat equation with periodic boundary conditions for `t` in `[0,1]` if everything is designed correctly:

``` julia
u0=Fun(θ->cos(cos(θ-0.1)),PeriodicInterval())
prob = ODEProblem((t,u) -> u'',u0,(0,1))   # ' is overriden for a Fun to mean derivative
```

(In practice, there should be a `chop` to make sure the length doesn't blow up, but that's a secondary consideration.)

**Comment by dlfivefifty** [comment]
`Fun` is `immutable`, but for no particular reason.

**Comment by dlfivefifty** [comment]
(My policy is to always use `immutable` to force me to think before I change to `type`.)

**Comment by ChrisRackauckas** [comment]
I think that immutable vs mutable is a better way of making the distinction. The reason is because we accept two different kind of functions: `du = f(t,u)` and `f(t,u,du)` which updates `du` in-place. The first doesn't require mutability. It seems like it should all work out if the types turn out all right. I don't know if this'll work with Sundials/ODEInterface though, I'm pretty sure they need arrays of `Float64`.

**Comment by dlfivefifty** [comment]
I think it's better to leave it immutable in this context: the number of coefficients of a  `Fun` can change with each time step, and so it's hard to see off the shelf code handling this.

**Comment by ChrisRackauckas** [comment]
> I think it's better to leave it immutable in this context: the number of coefficients of a Fun can change with each time step, and so it's hard to see off the shelf code handling this.

OrdinaryDiffEq.jl (the native DifferentialEquations.jl methods) actually already handle that.  See the cell model:

https://juliadiffeq.github.io/DiffEqDocs.jl/latest/features/callback_functions.html

**Comment by ChrisRackauckas** [comment]
But would they change size with every function call? That would be too difficult to handle.

**Comment by dlfivefifty** [comment]
Yes.   There is another way we could go about this: send only matrices and arrays to DiffEq.jl, and have this package pre and post process the data...

**Comment by ChrisRackauckas** [comment]
Would they still be changing size every function call? Because then it might be too difficult to handle anyways, and if you're always allocating to resize I don't think that there'd be too much of a performance difference trying to just chop a few of them out.

**Comment by ChrisRackauckas** [comment]
Question: what does broadcasting do to a `Fun`?

**Comment by MikaelSlevinsky** [comment]
Chris, something of interest for time evolution of PDEs is the use of IMEX Runge-Kutta schemes, i.e. schemes with multiple tableaus to handle multi-physics problems. These would work well with ApproxFun, because one can cache partial QR factorizations of linear operators (whose inverses are required in DIRK schemes for example). Should we consider these in DifferentialEquations.jl?

Here's a few IMEX RK schemes http://www.cs.ubc.ca/~ascher/papers/ars.pdf

**Comment by ChrisRackauckas** [comment]
These are already being considered, just haven't been implemented yet. It'll look pretty much the same as the ODEProblem:

``` julia
prob = ODEIMEXProblem(f,g,u0,tspan)
```

where you give the two functions, stiff and non-stiff, and then

``` julia
solve(prob,alg;kwargs...)
```

solves the problem with the chosen algorithm. That's the interface we're looking for. FYI I'm going to be concentrating on more Rosenbrock and exponential RK methods first though.

**Comment by ChrisRackauckas** [comment]
I'm going to take a crack at getting equations on `Fun` working on RK methods first. If I can get that going and have some tests on it, then I'll be able to plan accordingly for the other methods.

**Comment by ChrisRackauckas** [comment]
> These would work well with ApproxFun, because one can cache partial QR factorizations of linear operators (whose inverses are required in DIRK schemes for example).

Is that something that's done automatically? Can you show the docs for that?

**Comment by MikaelSlevinsky** [comment]
Ok. We have an ETDRK4 scheme built here somewhere. It uses some special functions from ApproxFun's `specialfunctions.jl` like `expα`, `expβ`, and `expγ` for numerical stability.

**Comment by MikaelSlevinsky** [comment]
I think you just call `qrfact` on an `Operator` and it stores the rest. Sheehan implemented this.

**Comment by dlfivefifty** [comment]
> Question: what does broadcasting do to a Fun?

``` julia
f.([1,2,3]) == [f(1),f(2),f(3)]

```

**Comment by dlfivefifty** [comment]
I'll add qrfact to the docs.

**Comment by dlfivefifty** [comment]
>  Question: what does broadcasting do to a Fun?

Sorry, maybe this is more helpful

``` julia
f=Fun(exp,[-1,1])
g=Fun(cos,[-1,1])
h=Fun(x->exp(cos(x)),[-1,1])
f.(g) == h  
```

**Comment by dlfivefifty** [comment]
Or also

``` julia
g=Fun(cos,[-1,1])
h=Fun(x->exp(cos(x)),[-1,1])
exp.(g) == h  
```

**Comment by dlfivefifty** [comment]
This code works for heat equation with periodic boundary conditions:

``` julia
S=Fourier()
u0=Fun(θ->cos(cos(θ-0.1)),S)
tspan = (0.,1.)
f(t,u) = (Fun(u,S)'').coefficients
prob = ODEProblem(f,u0.coefficients,tspan)
u = solve(prob)
plot(Fun(u(1.0),S))
```

**Comment by dlfivefifty** [comment]
I just pushed an update that supports the following code to solve `u_t = u_xx + (cos θ + 1) u_x`, by truncating at each stage to be the same number of coefficients as `u0`:

``` julia
using ApproxFun, DifferentialEquations, SpectralTimeStepping, Plots

S=Fourier()
u0=Fun(θ->cos(cos(θ-0.1)),S)
c=Fun(cos,S)

prob = SpectralTimeSteppingProblem((t,u)->u''+(c+1)*u',u0,(0.,1.))
@time u=solve(prob); #   4.087185 seconds (5.85 M allocations: 226.447 MB, 2.34% gc time)

@gif for t in linspace(0.,1.,100)
    plot(u(t);ylims=(0,1))
end
```

**Comment by dlfivefifty** [comment]
Changing the equation `(t,u)->0.001u''+(c+1)*u'` demonstrates the limitations of a fixed discretization size.  It can be improved by imposing extra coefficients for `u0`:

``` julia
S=Fourier()
u0=Fun(θ->cos(cos(θ-0.1)),S,200)
c=Fun(cos,S)

prob = SpectralTimeSteppingProblem((t,u)->0.001u''+(c+1)*u',u0,(0.,10.))
@time u=solve(prob); #   76.336780 seconds (239.30 M allocations: 9.011 GB, 4.10% gc time)
```

but the timing is not very impressive, probably due to CFL.  It would be good to incorporate implicit methods here.

**Comment by ChrisRackauckas** [comment]
> but the timing is not very impressive, probably due to CFL. It would be good to incorporate implicit methods here.

I set it up to work with the algorithm choices:

``` julia
@time u=solve(prob,CVODE_BDF());
```

That then uses Sundials' adaptive BDF method for 12 seconds, and 

``` julia
@time u=solve(prob,Trapezoid(),dt=1/4);
```

makes it do the Trapezoid (Crank-Nicholson) for 40 seconds (this is on OrdinaryDiffEq.jl master, there'll be a tag that solidifies that syntax). Our stiff native solvers are still being worked on, but this allows any new algorithm we make/optimize to be used. So this should be a good long-term strategy.

That said

> the limitations of a fixed discretization size. 

This is only a limitation because I did not have enough foresight. I knew something like this would come up, which is why I wanted to hammer out what it would be.

DifferentialEquations.jl offers two ways to declare a function: `f(t,u)` for `Number` and `f(t,u,du)` for `AbstractArray`. In general, most packages only support the later syntax (only OrdinaryDiffEq.jl's solver support the first), and what actually happens is the two-argument function is transformed into the three argument function. You should do this yourself for performance, but it's an ease of use thing.

However, the distinction from before was based off of `AbstractArray` and `Number`. So you probably tried to give a `Fun` directly, and it failed, so you made it use a fixed size array. Note that the first syntax should actually work as long as `+`, `-`, `*`, `/`, `abs` are defined for a `Fun`. So really, what's wrong is that I should've been dispatching on `mutable` vs `immutable` (`f(t,u)` treats all immutable types as scalars, `f(t,u,du)` handles any mutable). The mutable form will could then only change the sizes if `+`, `-`, `*`, `/` can deal with the different sizes.

Does that make sense? So I think I can play with it a bit and get it to work directly on changing discretization sizes. That will only make native Julia solvers compatible with it (which, if overtime our growth matches that of the non-stiff solvers, we should be getting better performance than other packages), so it would be nice to have the problem object offer the ability for fixed and changing size since the fixed size will (and already is) be compatible with any method, even those that are from C/Fortran.

**Comment by ChrisRackauckas** [comment]
While this won't be a priority for a bit (I have a few other things first, say a few months before I'd really get back to this), I'd like to start hacking away at this. Do you think this should stay in ApproxFun or should this (or something based off of it) go to JuliaDiffEq? I'd like to make some solid problem and solution types to match the full interfaces, and then a bunch of "pre-made PDEs" which are discretized via this method. Let me know what you think, but I'd like to at least get push access will likely work want to have things like a separate docs and stuff for it. I think we can really get a good piece of software out of this combo.

**Comment by dlfivefifty** [comment]
I gave you write access to this repos.  Let me know if you need access to ApproxFun.jl as well.

**Comment by dlfivefifty** [comment]
I also had a go overriding to get it to work with Fun instead of Number by overriding routines:

``` julia
RecursiveArrayTools.recursive_one(f::Fun) = ones(space(f))
OrdinaryDiffEq.ODE_DEFAULT_NORM(f::Fun) = OrdinaryDiffEq.ODE_DEFAULT_NORM(f.coefficients)
OrdinaryDiffEq.ODE_DEFAULT_NORM{F<:Fun}(f::Array{F}) = 
OrdinaryDiffEq.ODE_DEFAULT_NORM(map(OrdinaryDiffEq.ODE_DEFAULT_NORM,f))
```

but I got stuck pretty quickly.

**Comment by dlfivefifty** [comment]
> Do you think this should stay in ApproxFun or should this (or something based off of it) go to JuliaDiffEq?

No preference.  I think the current plan is to rename the organization JuliaApproximation so that it can be a more general home.

**Comment by ChrisRackauckas** [comment]
> I also had a go overriding to get it to work with Fun instead of Number by overriding routines:

Yeah, there's a few things there that will probably need to be worked out:
1. Dispatch `Fun` where I was dispatching `Number` before (probably changing to traits).
2. The right error norm on a `Fun`
3. Getting the `Vector{Fun}` part of the solution to "work right" with interpolations when sizes are changing.
4. Getting the right types for the internal values. The current setup will try to make `abstol` and `reltol` be `Fun`, which I can see causing problems.

But in the end making it work this way probably will be much better (for the algorithms that can handle it).

> No preference. I think the current plan is to rename the organization JuliaApproximation so that it can be a more general home.

Then I think what I'd like to do is make an ApproxFunDiffEq.jl in JuliaDiffEq which will be just for making ApproxFun types compatible with DifferentialEquations.jl and the component solvers. From this small amount of playing around, it looks like that work is pretty much done on your side, but I'll have some finagling around to get the interaction up to speed, so most of the code will probably be DiffEq specific. I think ApproxFunDiffEq.jl is a good clear name for the interaction (matches both naming schemes). And if it's in JuliaDiffEq, you won't get mad when I fill it up with a bunch of really JuliaDiffEq specific stuff 👍 .

I'll have to learn a bit more about ApproxFun to really make this go though, so it'll take some time.

My last question is then, is there anything that your timestepping methods like `ETDRK4` is doing that is "ApproxFun-specific"? Any optimizations which can only be done on `Fun` types which I should know about? I'm not quite sure what you're doing with the `cache` and `chop` operations.

**Comment by ChrisRackauckas** [comment]
BTW, I tried:

``` julia
@time u=solve(prob,radau());
@time u=solve(prob,rk45());
```

as well, so I can confirm this (fixed discretization size) method works with ODE.jl and ODEInterface.jl too through the DifferentialEquations.jl common interface.

**Comment by dlfivefifty** [comment]
The big question is how best to incorporate boundary conditions.  Usually I would incorporate them into an implicit linear time step, but I don't see how that fits into JuliaDiffEq.jl.  

For fixed discretizations, I suppose it could be accomplished as a DAE....

**Comment by ChrisRackauckas** [comment]
Isn't it usually imposed by the discretization? I know for FEM/FDM it's done by the discretization of the operators, and for periodic it's done by the choice of basis.

**Comment by dlfivefifty** [comment]
For spectral methods its usually imposed by extra constraints on the differential equation.  For example, consider the ODE `u(±1) = 0, u'' + u = f`.  The operator will have two extra rows corresponding to evaluation:

``` julia
S=Chebyshev()
D=Derivative(S)
A=[Evaluation(S,-1);
      Evaluation(S,1);
      D^2 + I]
```

and so the resulting operator has the form:

```
 1.0  -1.0   1.0                  -1.0    1.0                  …   1.0                  -1.0     ⋯
 1.0   1.0   1.0                   1.0    1.0                      1.0                   1.0     ⋯
 1.0   0.0   3.3333333333333335    0.0    0.16666666666666666                                     
       0.25  0.0                   5.625  0.0                                                     
             0.16666666666666666   0.0    7.733333333333333                                       
                                   0.125  0.0                  …                                  
                                          0.1                      0.07142857142857142            
                                                                   0.0                   0.0625   
                                                                  15.873015873015873     0.0     ⋱
                                                                   0.0                  17.8875  ⋱
                                                               …    ⋱                     ⋱      ⋱
```

where the first two rows are boundary rows.  The constraints are then imposed on the right-hand side, so one would solve:

``` julia
u=A\[0.;0.;f]
```

so that `Evaluation(S,-1)*u ≈ 0`, `Evaluation(S,1)*u ≈ 0` and `(D^2+I)*u ≈ f`.

In the context of PDEs, suppose we solve heat equation with Dirichlet: `u_t = u_xx, u(t,±1) = 0, u(0,x) = u0`.  If we use backward Euler, we include the extra constaints in the implicit solve:

``` julia
h=0.01
A=[Evaluation(S,-1);
      Evaluation(S,1);
      I-h*D^2]

u[1] = u0
u[k+1] = A\[0.;0.;u[k]]
```

I'm not sure how to fit this into DifferentialEquations.jl

**Comment by dlfivefifty** [comment]
Note that there is some auto-differentiation support implemented for nonlinear solvers, which could be used to suss out the structure from an anonymous function, e.g. heat equation with Dirichlet could be specified via

``` julia
(t,u,du) -> [u(-1);u(1);du-u'']
```

**Comment by mauro3** [comment]
Is this then not a Differential Algebraic equation (DAE)?  This is supported by the Sundials IDA and the DASSL.jl solvers.

**Comment by ChrisRackauckas** [comment]
> Is this then not a Differential Algebraic equation (DAE)? This is supported by the Sundials IDA and the DASSL.jl solvers.

Yeah that's probably overkill though and will limit the methods which can be used.

> we include the extra constaints in the implicit solve

Are you doing this in timeevolution.jl? If so, how so? Do you just need to be able to inject a function somewhere?

**Comment by dlfivefifty** [comment]
Yes I think so, for fixed discretizations.

Ideally one could leveraging the fact that \ is O(N), which will be lost in a general DAE solver I think, though that might be too hard.

**Comment by dlfivefifty** [comment]
In `BDF2` on line 15, the argument `B` is a vector of functional constraints, with `bcs` their values.  So in the example above, one could take

``` julia
B=[Evaluation(S,-1);
      Evaluation(S,1)]
bcs=[0.;0.]
```

`SBDF2` is then the operator including the boundary conditions.  `g` contains the nonlinear part, which is handled explicitly using RK4.   `glp` was to support live plotting with Reactive.jl.

**Comment by dlfivefifty** [comment]
I'm working on the docs a bit:

http://approxfun.github.io/ApproxFun.jl/latest/

**Comment by ChrisRackauckas** [comment]
> SBDF2 is then the operator including the boundary conditions. g contains the nonlinear part

Wouldn't these two things be what you would use to make the DiffEq `f`? Put the bcs into `A` and then `f = A*u+g`. And solve that `f` using whatever method? (We don't have IMEX methods yet, but when we do you would just give the two parts). Maybe I need to play with ApproxFun to see why this could/couldn't be the case.

**Comment by dlfivefifty** [comment]
The issue is that boundary conditions (other than periodic) are not preserved: just because `u(±1) = 0` does not mean `u''(±1) = 0`.

The fact time-stepping works well with periodic boundary conditions is extremely lucky.

**Comment by dlfivefifty** [comment]
But we might be getting ahead of ourselves: I'm not sure boundary conditions are realistic without implicit-explicit (IMEX?) methods.

IMEX are mostly useful when you have fast implicit solvers, so for that we'd need to specify to use `qrfact` for the implicit part.

**Comment by ChrisRackauckas** [comment]
>I'm not sure boundary conditions are realistic without implicit-explicit (IMEX?) methods.

They should be able to be imposed on the explicit method as well. For spectral methods it's usually through the choice of basis (periodic -> Fourier basis). If there is no appropriate basis, then there is an order reduction if you impose the restriction directly on ODEs.

When the appropriate basis can't be formed, then a simplified form of a DAE is usually sufficient for this. I think what we discussed can always be a constrained ODE or an ODE with a mass matrix. Getting these right (without simply making everything an implicit ODE) will take some time. But I think we know what we need to do now.

>so for that we'd need to specify to use qrfact for the implicit part.

What is it just `qrfact`? For many cases in ODEs the LU factorization is used. Does that not make sense for `Fun`'s (is it non-square?)

**Comment by dlfivefifty** [comment]
> When the appropriate basis can't be formed, then a simplified form of a DAE is usually sufficient for this. 

I agree that this is the simplest way forward to handle general boundary conditions, though you lose the O(N) complexity of ApproxFun's \ as it won't recognize that the operators are almost banded.

> What is it just qrfact? For many cases in ODEs the LU factorization is used. Does that not make sense for Fun's (is it non-square?)

Short answer: only `qrfact` has been implemented.  The operators are infinite-dimensional (the factorization is calculated adaptively), so "square" is not the right word.  An implementation of `lufact` is easy without pivoting, however, this is prone to instability hence the default is `qrfact`.   I don't think `lufact` with pivoting without giving up the O(N) complexity as they are banded operators.

The infinite-dimensional nature of ApproxFun works well with implicit-explicit methods, which extend naturally to ∞-dimensions.  The benefit of building off of ApproxFun's `qrfact` is that each time step will be automatically resolved adaptively to a prescribed tolerance, so you get spacial adaptivity for "free" (there are some catches that you have to be careful about L-stability..).    Do you have a reference for order reduction so I can see if this is compatible?


PS Here's a spectral time stepping project in python:

http://dedalus-project.org

I believe they work using implicit-explicit methods where the user specifies the equation as
```
L u_t + M u = f(t,u)
```
where `L` and `M` are matrices.  The way boundary conditions are imposed is that `L` can be singular, so you have say (using ApproxFun terminology)
```julia
L  = [0; 0; I]
M = [Evaluation(-1);Evaluation(1);-D^2]
f = (t,u) -> [0;0;u^2]
```
The departure here from DifferentialEquations.jl is that it allows a matrix acting on the `u_t` term.

**Comment by ChrisRackauckas** [comment]
>The departure here from DifferentialEquations.jl is that it allows a matrix acting on the u_t term.

That's mass matrices. We already know how we want to do that. https://github.com/JuliaDiffEq/DiffEqBase.jl/issues/1 . And lots of solvers will be able to handle that with only minor changes (i.e. just expose the API to pass the mass matrix). However, it'll take some time to implement since I'll be working on stochastic things more for awhile.

**Comment by ChrisRackauckas** [comment]
>Short answer: only qrfact has been implemented. The operators are infinite-dimensional (the factorization is calculated adaptively), so "square" is not the right word. An implementation of lufact is easy without pivoting, however, this is prone to instability hence the default is qrfact. I don't think lufact with pivoting without giving up the O(N) complexity as they are banded operators.

Got it. We'll just need to expose the choice of factorization method, and set defaults when using `Fun`s.

**Comment by dlfivefifty** [comment]
Yep, I think that'll work.  Then code that treats `L` and `M` as matrices should hopefully work when they are swapped out for operators.  

This should also resolve another issue: `rangespace(M) ≠ space(u_t)`, so you need a conversion operator.  That is, `u_t + M*u` becomes  (pseudo-code here)

```julia
S=space(u0)
C=Conversion(S,rangespace(M))
C*u_t + M*u
```

**Comment by ChrisRackauckas** [comment]
>Then code that treats L and M as matrices should hopefully work when they are swapped out for operators.

I currently use it with the `I` (UniformScaling) operator, so I assume other operators will "just work" if they have the same actions. 

>This should also resolve another issue: rangespace(M) ≠ space(u_t), so you need a conversion operator. 

That conversion should just be applied before making the mass matrix problem and passed as the mass matrix, right? If it needs to be applied each step (to deal with the changing discretization size? Or will it handle it automatically?), then the non-constant mass matrix version will allow it to be a function which will be able to do what's needed here.

**Comment by dlfivefifty** [comment]
Yep, the conversion operator would be the mass matrix.  Operators can handle the adaptivity automatically (though a call to `cache(C)` will adaptively cache the entries of the operator, which will usually improve the speed of multiplication).

Would it be helpful to have `A_mul_B!`, etc., implemented for Funs and Operators?

**Comment by ChrisRackauckas** [comment]
I think so? What does `A_mul_B!` in your case though? I thought your `Fun`s were immutable?

For the "can be indexed" case I use InPlaceOpts.jl to make everything in-place (and thus use `A_mul_B!`), but I don't think `Fun`s should be used/indexed like arrays, right?

**Comment by dlfivefifty** [comment]
Ah but the coefficients are (mutable) vectors, so one could do, for example:
```julia
resize!(f.coefficients,1000)
f.coefficients[:]=rand(1000)
```

It is possible to make a Fun usable as an ∞-dimensional array, via:
```julia
c=Fun(f.coefficients,SequenceSpace())
c[10] # returns 10th coefficient
```
But I'm not sure why a time-stepper would need to know about indexing.


PS I'm about to change the order to `Fun(space,coefficients)` so that its semantically different from `Fun(function,space)`, so the code above would need to be altered.

**Comment by ChrisRackauckas** [comment]
>But I'm not sure why a time-stepper would need to know about indexing.

That's because in the most common case you update each index at each step. Until all of the broadcast fusing changes are done, `for i in eachindex(u) ... end` is the best way to do this. Afterward, linear combinations will be able to be efficiently done with `utmp .= a31*k[1] .+ a32*k[2]`. Would `Fun`'s of different "sizes" (numbers of coefficients) still broadcast fine? And in-place update fine like that?

**Comment by dlfivefifty** [comment]
Broadcasting should be fine.  I could overload `.=` to do the right thing:

```julia
function Base.broadcast!(::typeof(identity),a::Fun,b::Fun)
   resize!(a.coefficients,ncoefficients(b))
   a.coefficients[:]=b.coefficients
   a
end
``

**Comment by ChrisRackauckas** [comment]
Cool. Then it should all workout come v0.6.

**Comment by dlfivefifty** [comment]
I guess to support fusing properly I probably want to also have the definition:
```julia
function Base.broadcast!(f,a::Fun,b::Fun)
    c=f.(b)
    a.=c
end
```

**Comment by ChrisRackauckas** [comment]
To fuse what statement?

**Comment by dlfivefifty** [comment]
I'm worried about autofusing: if I have `a .= f.(b)` won't it try to fuse the operations, which will miss the `Base.broadcast!(::typeof(identity),a::Fun,b::Fun)` override?

**Comment by ChrisRackauckas** [comment]
That is true. I think you do need the extra definition.

**Comment by dlfivefifty** [comment]
@ChrisRackauckas I've added documentation to ApproxFun: http://juliaapproximation.github.io/ApproxFun.jl/stable/

Hopefully this clarifies the relationship between operators, qrfact, etc. a bitl.

If you have any questions/requests just file an issue.  One major change is that the syntax is now `Fun(exp,0..1)` instead of `Fun(exp,[0,1])` for creating a function on the interval [0,1].

**Comment by ChrisRackauckas** [comment]
Using QR / allowing the choice of QR won't be a problem if the linear algebra interface I'm interested in becomes a thing. See the blog post:

http://www.stochasticlifestyle.com/modular-algorithms-scientific-computing-julia/

With that, someone could just pass in a method, so I'd default to LU, but could make the dispatch on `Fun` use QR (as a default).

**Comment by MikaelSlevinsky** [comment]
Nice article! P.S. I think your linear solves are backwards (or maybe it's just my version of Julia that's missing new methods):
```julia
julia> A = rand(3,3)
3×3 Array{Float64,2}:
 0.628874  0.527016  0.465296
 0.151401  0.401624  0.197612
 0.18347   0.674264  0.4713  

julia> b = rand(3)
3-element Array{Float64,1}:
 0.979039
 0.348729
 0.747334

julia> K = lufact(A)
Base.LinAlg.LU{Float64,Array{Float64,2}}([0.628874 0.527016 0.465296; 0.291743 0.52051 0.335553; 0.24075 0.527839 -0.0915263],[1,3,3],0)

julia> x = b\K
ERROR: ctranspose not implemented for Base.LinAlg.LU{Float64,Array{Float64,2}}
 in \(::Array{Float64,1}, ::Base.LinAlg.LU{Float64,Array{Float64,2}}) at ./operators.jl:145

julia> x = K\b
3-element Array{Float64,1}:
  0.528412 
 -0.0334175
  1.42779  

```

**Comment by ChrisRackauckas** [comment]
haha I was silly. Thanks for noticing that.

**Comment by dlfivefifty** [comment]
Looks good!  You can also use `factorize` as the default, which is overloaded to use `qrfact` for operators

**Comment by ChrisRackauckas** [comment]
Oh, that might be a good idea. But `factorize` isn't type-stable, right?

**Comment by dlfivefifty** [comment]
I suppose not, but type stability of a factorization object is unlikely to cause overhead, as it immediately dispatches to a \ call.

**Comment by ChrisRackauckas** [comment]
Ahh, but I will want to store the factorized object in many cases because it can be re-used. I though `factorize` returned the factorized matrix? I forget.

**Comment by ChrisRackauckas** [comment]
Just updating this thread.

The discussion over here solves the matrix factorization problem we were talking about. The factorization function could just be what's passed in, and default to factorize, and this should cover all linear solvers in the future as well.

https://discourse.julialang.org/t/unified-interface-for-linear-solving/699

Is there rootfinding on `ApproxFun` types? What would it even mean? That would be necessary for things like BDF methods. But whatever it is, if we unify the rootfinding interface as well, it could be overloaded for `ApproxFun` types and seamlessly work in those kind of methods:

https://discourse.julialang.org/t/a-unified-interface-for-rootfinding/698

That'll be what's needed for implicit methods on the adaptive-basis `Fun` to work.

**Comment by dlfivefifty** [comment]
There's a routine `newton` that does root finding for ODEs, usually BVPs:

```julia
x=Fun() 
u0=0.0x  # initial guess
N=u->[u(-1)-1,u(1)+0.5,0.001u''+6*(1-x^2)*u'+u^2-1] u=newton(N,u0)
```

It would be awesome to use a general root finding package.  In this setting, Jacobians are `Operator`s, which I calculate using auto differentiation via a type `DualFun`
